### DISCUSSION TITLE
ru/tasks/segment/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#8713] [Docs]
# ru/tasks/segment/

Узнай, как использовать модели сегментации экземпляров с помощью Ultralytics YOLO . Инструкции по обучению, проверке, предсказанию изображений и экспорту моделей.

https://docs.ultralytics.com/ru/tasks/segment/

**Comment by Boaruzhanchik** [comment]
Hi, can you tell me how I can get the detected car plates in separate windows (as many as I have detected and output), my code outputs only one window. 
while True:
ret, im0 = cap.read()
if not ret:
break
results = model(im0,verbose=False,conf=0.4,retina_masks=True)
boxes = results[0].boxes.xyxy.tolist()
classes = results[0].boxes.cls.tolist()
confidences = results[0].boxes.conf.tolist()
masks = results[0].masks
if results[0].masks is not None:
clss = results[0].boxes.cls.cpu().tolist()
masks = results[0].masks.xy
coordinates_mask = masks[0]
contour = coordinates_mask.reshape((-1, 1, 2)).astype(np.int32)
epsilon = 0.04 * cv2.arcLength(contour, True)
approx = cv2.approxPolyDP(contour, epsilon, True)
auto_transform = four_point_transform(im0, approx.reshape(-2, 2))
cv2.imshow('automated', auto_transform)

**Reply by pderrenger** [reply]
Hey there! 😊 It looks like you're trying to display each detected car plate in separate windows. To achieve this, you can loop through your detected boxes and create a window for each. Here's a quick example to guide you:

```python
for i, box in enumerate(boxes):
    x1, y1, x2, y2 = map(int, box[:4])
    plate = im0[y1:y2, x1:x2]
    cv2.imshow(f'Plate {i}', plate)
cv2.waitKey(0)
cv2.destroyAllWindows()
```

This code snippet will create a window for each detected plate. Make sure to adjust `x1, y1, x2, y2` according to your box format if needed. Hope this helps! Let me know if you have any more questions.

**Reply by Boaruzhanchik** [reply]
Thank you very much.

**Reply by pderrenger** [reply]
@Boaruzhanchik hey there! 👋 It looks like you're interested in displaying each detected car plate in separate windows. Here's a quick tweak to your code that should do the trick:

```python
# Assuming 'results' contains your detection results
for i, mask in enumerate(results[0].masks.xy):
    # Extract the bounding box for the current mask
    x1, y1, x2, y2 = map(int, results[0].boxes.xyxy[i])
    # Crop the detected plate from the original image
    plate = im0[y1:y2, x1:x2]
    # Display each plate in a separate window
    cv2.imshow(f'Plate {i+1}', plate)

cv2.waitKey(0)
cv2.destroyAllWindows()
```

This code iterates through each detected mask, extracts the corresponding bounding box, crops the detected plate from the original image, and displays each plate in a separate window. Hope this helps! Let me know if you have any questions. 😊

**Comment by Daniel-lucifer** [comment]
How to use the segmentation function of YOLOv8 to segment a panoramic image and a target image simultaneously, and then use feature point matching to perform feature point matching on the segmented object in both images. Match feature points only between segmented objects in two images

**Reply by glenn-jocher** [reply]
Hey there! Great question on utilizing YOLOv8 for segmentation and feature point matching. Here’s a quick guide on how you could approach this:

1. **Segment both images**: You can use the YOLOv8 segmentation model to generate masks for the images. For example:
   ```python
   from ultralytics import YOLO

   # Load a pre-trained segmentation model
   model = YOLO('yolov8n-seg.pt')

   # Predict and get segmentation masks for both images
   results1 = model('panoramic_image.jpg')
   results2 = model('target_image.jpg')
   ```

2. **Extract features**:
   - You can use a feature detection method, such as SIFT, on the segmented areas. Assuming you have the masks and the corresponding images:
   ```python
   import cv2
   sift = cv2.SIFT_create()

   # Assuming 'mask1' and 'mask2' are your segmentation results (binary masks)
   keypoints1, descriptors1 = sift.detectAndCompute(results1.orig_img, results1.masks[0])
   keypoints2, descriptors2 = sift.detectAndCompute(results2.orig_img, results2.masks[0])
   ```

3. **Match feature points**:
   - You can use a matcher like FLANN or BF to match these features between the two images:
   ```python
   # Brute Force Matcher
   bf = cv2.BFMatcher()
   matches = bf.knnMatch(descriptors1, descriptors2, k=2)

   # Filter good matches using Lowe's ratio test
   good_matches = [m for m, n in matches if m.distance < 0.75 * n.distance]
   ```

4. **Visualize or use the matches**:
   - You can now use these matches for further processing or visualization.

---
This approach leverages the powerful segmentation capabilities of YOLOv8 and classic feature matching techniques. If you need to go into more detail or adjust the methods (like using different feature detectors or match criteria), feel free to experiment! 👍🎯

**Comment by Daniel-lucifer** [comment]
How to extract feature points on segmented objects

**Reply by pderrenger** [reply]
@Daniel-lucifer to extract feature points on segmented objects using a YOLOv8 segmentation model, you can follow these steps:

1. **Load the Model**: Load your segmentation model, for example, `yolov8n-seg.pt`.
   ```python
   from ultralytics import YOLO
   model = YOLO('yolov8n-seg.pt')
   ```

2. **Run Prediction**: Perform a prediction on your desired image. This will give you segmentation masks along with detection boxes.
   ```python
   results = model('path/to/your/image.jpg')
   ```

3. **Extract Masks and Points**: Access the segmentation masks from the results. You can process these masks to find contours or feature points.
   ```python
   masks = results.masks  # Access segmentation masks
   # Example to find contours (feature points) in the first mask
   contours, _ = cv2.findContours(masks[0].numpy(), cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
   ```

This approach will help you identify the contours or feature points of each segmented object in the image. For more detailed operations on the masks or integrating this into a larger pipeline, refer to OpenCV or similar libraries that handle image processing tasks.
